prov/verbs/src/verbs_ep.c

```diff
@@ -1034,15 +1034,19 @@ static int vrb_ep_enable(struct fid_ep *ep_fid)
 			return -FI_EINVAL;
 		}
 
-		ret = rdma_create_qp(ep->id, domain->pd, &attr);
-		if (ret) {
-			VRB_WARN_ERRNO(FI_LOG_EP_CTRL, "rdma_create_qp");
-			return -errno;
-		}
+		/* Server-side QP creation, after RDMA_CM_EVENT_CONNECT_REQUEST
+		 * is recevied */
+		if (ep->id->verbs) {
```

Comment:
I'm still having a hard time with this one.  Using internal variable from another library for this purpose is just a bad idea.  Unfortunately, I do not have an alternative for you.

Reply:
The verbs provider already has a couple of existing conditions based on `id->verbs`. For example:
https://github.com/ofiwg/libfabric/blob/main/prov/verbs/src/verbs_info.c#L1316
https://github.com/ofiwg/libfabric/blob/main/prov/verbs/src/verbs_info.c#L1789

There is even this comment in the code:
```
	/* Handle the case when rdma_cm doesn't fill src address even
	 * though it fills the destination address (presence of id->verbs
	 * corresponds to a valid dest addr) */
```
See https://github.com/ofiwg/libfabric/blob/main/prov/verbs/src/verbs_info.c#L1319C3-L1321C41

I'm not adding something new to the code.

Reply:
I don't have that kind of insight into libibverbs nor librdmacm and I certainly would not code that way.  Just me.

Reply:
id->verbs is externally visible.  It has to be that way to coordinate the use of librdmacm with libibverbs.  struct rdma_cm_id has both visible and hidden fields.  The hidden fields are not accessible to a user.

Reply:
`struct rdma_cm_id` is part of the public API of RDMA CM. Referencing its member is fine when needed.

Reply:
It is one thing to use the value, it is another to infer behavior.  Can we be certain when id->verbs is set, it is always after RDMA_CM_EVENT_CONNECT_REQUEST is received on the server side?

Reply:
If we rephrase the question to: Is id->verbs always set for an RDMA_CM_EVENT_CONNECT_REQUEST?  The answer is yes.  Connect requests are received over a specific device.

Chien's question is subtly different, however, and I think the answer to his question is *no*, which might be exposing an issue.

An rdma_cm_id is associated with a verbs device in 3 cases.  On the client side, this is the purpose of rdma_resolve_addr().  Once that operation completes, we have a device.  On the server side, all connection requests are associated with a device.  The 3rd case, however, is through rdma_bind_addr().

The restructured connect() path calls enable() first, then rdma_resolve_addr().  If we're taking the client code path, then id->verbs is *usually* not set.  However, if we went through rdma_bind_addr() prior to calling connect(), it may be.  In that case, I think we would end up trying to create 2 QPs.

Reply:
Sorry for the very, very late reply :)

I updated the code to check that ep->ibv_qp is NULL before creating the QP. The change is available here:
https://github.com/ofiwg/libfabric/pull/9543/files#diff-0c780e52c4b4880ab4a4012f359d1ae818bbdddcd650d9192f250f622e7bf927R1039

FYI, our product has been running with this PR for the last 2 months with no issue: plain Verbs (no RXM) and MSG endpoints. The only real issue we had was with the following patch that introduces EP states to the Verbs provider: https://github.com/ofiwg/libfabric/pull/9175
The above PR was causing assertion failures during connection shutdown because the EP state was incorrect. We simply reverted this patch and everything works now.

Anyway, I'm still convinced this PR is important for upstream as some runtime models require all the calls to be non-blocking (I believe this is the case for DAOS too because of the SPDK model). Please review it when you can.